src/stpipe/library.py

```diff
@@ -0,0 +1,850 @@
+        for i in range(len(self)):
+            yield self.borrow(i)
+
+    def _assign_member_to_model(self, model, member):
+        """
+        Assign association member information to an opened model.
+
+        This will be called:
+
+            - when a model is loaded (for the first time) in ``_load_member``
+            - when the library is created (if the library was provided a
+              list of open models)
+
+        This will assign the following metadata attributes (setting
+        for example ``model.meta.group_id`` to the value of
+        ``member["group_id"]``):
+
+            - group_id
+            - exptype
+            - tweakreg_catalog
+
+        and (when available from the association) assign the following
+        metadata attributes based on the association information:
+
+            - set ``meta.asn.table_name`` to ``asn["table_name"]``
+            - set ``meta.asn.pool_name`` to ``asn["pool_name"]``
+
+        Parameters
+        ----------
+        model : DataModel
+            Model to be updated (in place).
+
+        member : dict
+            Dictionary containing contents of the "member" entry
+            (include "exptype", "expname", "group_id", etc...)
+        """
+        for attr in ("group_id", "tweakreg_catalog", "exptype"):
+            if attr in member:
+                setattr(model.meta, attr, member[attr])
+
+        if not hasattr(model.meta, "asn"):
+            model.meta.asn = {}
+
+        model.meta.asn["table_name"] = self.asn.get("table_name", "")
+        model.meta.asn["pool_name"] = self.asn.get("asn_pool", "")
+
+    def _load_member(self, index):
+        """
+        Load a model for the association member at the provided index.
+
+        This will be called when the model at the provided index:
+            - is not already loaded
+            - does not have a temporary file
+
+        Parameters
+        ----------
+        index : int
+
+        Returns
+        -------
+        model : DataModel
+        """
+        member = self._members[index]
+        filename = os.path.join(self._asn_dir, member["expname"])
+
+        model = self._datamodels_open(filename, **self._datamodels_open_kwargs)
+
+        self._assign_member_to_model(model, member)
+
+        return model
+
+    def save(self, path, **kwargs):
+        """
+        .. warning:: This save is NOT used by Step/Pipeline. This is
+                     intentional as the Step/Pipeline has special requirements.
+
+        For now this is a very basic "save". It does not:
+
+            - check that the library contains no duplicate filenames
+            - propagate asn_pool and other non-member asn information
+
+        It will save:
+            - models with filenames determined by ``_model_to_filename``
+            - a very simple association that contains only 1 product that
+              lists ``expname`` ``exptype`` and ``group_id``  for each model
+              in a file named "asn.json".
```

Comment:
Would it be possible/helpful to allow user-specified filenames?

Reply:
I think that's covered by setting the `model.meta.filename` (or whatever is used for `_model_to_filename`) is set prior to the `save` call. I'll make this a private method (since it's not intended for use by the pipeline code).

Reply:
I renamed `save` to `_save` here: 1a24e7816f11563ab987c5c65d82d194c2f3db07

This does make `ModelLibrary` no longer a subclass of `AbstractDataModel`. That is fine it just meant the smoke test for this relationship had to change as well.